### PR TITLE
MSTR-14 Problem API 개발

### PR DESCRIPTION
**Pull request by kshired**
- Problem API 1차 개발 완료
  - 추후 권한 인증 작업등의 detail한 작업 예정

### CODE REVIEW

File: src/main/kotlin/com/csbroker/apiserver/common/auth/AuthToken.kt [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+import io.jsonwebtoken.UnsupportedJwtException
+import java.security.Key
+import java.util.Date
+import java.util.UUID
+
+const val AUTHORITIES_KEY: String = "ROLE"
+
+class AuthToken(
+    private var token: String,
+    private val key: Key,
```
Comment (kshired) [inline src/main/kotlin/com/csbroker/apiserver/common/auth/AuthToken.kt:19]: 🚫 **[ktlint]** <trailing-comma> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Unnecessary trailing comma before ")"

File: src/main/kotlin/com/csbroker/apiserver/common/auth/AuthTokenProvider.kt [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+import io.jsonwebtoken.security.Keys
+import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
+import org.springframework.security.core.Authentication
+import org.springframework.security.core.authority.SimpleGrantedAuthority
+import org.springframework.security.core.userdetails.User
+import java.security.Key
+import java.util.Date
+import java.util.UUID
+
+
```
Comment (kshired) [inline src/main/kotlin/com/csbroker/apiserver/common/auth/AuthTokenProvider.kt:13]: 🚫 **[ktlint]** <no-consecutive-blank-lines> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Needless blank line(s)

File: src/main/kotlin/com/csbroker/apiserver/common/auth/UserPrincipal.kt [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+import org.springframework.security.oauth2.core.oidc.user.OidcUser
+import org.springframework.security.oauth2.core.user.OAuth2User
+import java.util.Collections
+
+class UserPrincipal(
+    private val userId: String,
+    private val password: String,
+    private val providerType: ProviderType,
+    private val roleType: Role,
+    private val authorities: MutableCollection<GrantedAuthority>,
```
Comment (kshired) [inline src/main/kotlin/com/csbroker/apiserver/common/auth/UserPrincipal.kt:19]: 🚫 **[ktlint]** <trailing-comma> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Unnecessary trailing comma before ")"

File: src/main/kotlin/com/csbroker/apiserver/dto/ProblemSearchDto.kt [hunk 4]
```diff
@@ -0,0 +1,7 @@
+package com.csbroker.apiserver.dto
+
+data class ProblemSearchDto(
+    val tags: List<String> = listOf(),
+    val solvedBy: String? = null,
+    val query: String = "",
```
Comment (kshired) [inline src/main/kotlin/com/csbroker/apiserver/dto/ProblemSearchDto.kt:6]: 🚫 **[ktlint]** <trailing-comma> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Unnecessary trailing comma before ")"

File: src/main/kotlin/com/csbroker/apiserver/model/Problem.kt [hunk 5]
```diff
@@ -34,4 +36,56 @@ class Problem(
 
     @OneToMany(mappedBy = "problem")
     val problemTags: MutableList<ProblemTag> = mutableListOf(),
-) : BaseEntity()
+
+    @OneToMany(mappedBy = "problem")
+    val gradingHistory: MutableList<GradingHistory> = mutableListOf()
+) : BaseEntity() {
+    fun toProblemResponseDto(): ProblemResponseDto{
```
Comment (kshired) [inline src/main/kotlin/com/csbroker/apiserver/model/Problem.kt:43]: 🚫 **[ktlint]** <curly-spacing> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Missing spacing before "{"

File: src/main/kotlin/com/csbroker/apiserver/model/Problem.kt [hunk 6]
```diff
@@ -34,4 +36,56 @@ class Problem(
...
+
+        return ProblemResponseDto(
+            this.title,
+            tags,
+            avgScore,
+            totalSolved
+        )
+    }
+
+    fun toProblemDetailResponseDto(): ProblemDetailResponseDto{
```
Comment (kshired) [inline src/main/kotlin/com/csbroker/apiserver/model/Problem.kt:66]: 🚫 **[ktlint]** <curly-spacing> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Missing spacing before "{"

**Review by Kim-Hyunjo — APPROVED** [review]
👍

**Review by ekzm8523 — APPROVED** [review]
LGTM..⭐️

**Pull request merged**